### PR TITLE
Add MainThread.SetCustomImplementation for custom platform backends

### PR DESCRIPTION
**Pull request by Redth**
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

## Description

Fixes #34101

On custom platform backends (e.g. Linux/GTK), `MainThread.BeginInvokeOnMainThread()` and `MainThread.InvokeOnMainThreadAsync()` throw `NotImplementedInReferenceAssemblyException` because there is no platform-specific partial class implementation for the netstandard TFM.

This PR adds a public `MainThread.SetCustomImplementation()` method that allows custom backends to register their own `IsMainThread` and `BeginInvokeOnMainThread` implementations.

## Changes

- **`MainThread.shared.cs`**: Added static callback fields and `SetCustomImplementation(Func<bool>, Action<Action>)` public method
- **`MainThread.netstandard.cs`**: Modified to check for custom callbacks before throwing `NotImplementedInReferenceAssemblyException`
- **`PublicAPI.Unshipped.txt`**: Added new public API entry for all targets
- **`MainThread_Tests.cs`**: Added 6 unit tests covering the new functionality

## Usage

Custom platform backends can register their implementation during startup:

```csharp
MainThread.SetCustomImplementation(
    isMainThread: () => /* check if on UI thread */,
    beginInvokeOnMainThread: action => /* dispatch to UI thread */);
```

For example, a GTK backend might use:

```csharp
MainThread.SetCustomImplementation(
    isMainThread: () => GLib.MainContext.Default.IsOwner,
    beginInvokeOnMainThread: action => GLib.Idle.Add(() => { action(); return false; }));
```

On standard platforms (iOS, Android, Windows, MacCatalyst), the existing platform-specific implementations continue to be used unchanged.

### CODE REVIEW

File: src/Essentials/src/MainThread/MainThread.shared.cs [hunk 1]
```diff
@@ -9,6 +9,22 @@ namespace Microsoft.Maui.ApplicationModel
 	/// </summary>
 	public static partial class MainThread
 	{
+		static Func<bool> s_isMainThreadImplementation;
```
Comment (rmarinho) [inline src/Essentials/src/MainThread/MainThread.shared.cs]: do we use `s_` ?

**Comment by Redth** [comment]
Addressed today’s 2 new Copilot comments and verified with build/tests.

✅ Fixed
- `src/Essentials/src/MainThread/MainThread.shared.cs`
  - `SetCustomImplementation` now throws `PlatformNotSupportedException` on platform TFMs (`ANDROID`, `IOS`, `MACCATALYST`, `TVOS`, `WATCHOS`, `WINDOWS`, `TIZEN`) instead of silently storing delegates that are never used there.
  - Updated XML remarks to reflect that platform TFMs throw and only non-platform TFMs (e.g. netstandard/bare net) use custom callbacks.

- `src/Essentials/test/UnitTests/MainThread_Tests.cs`
  - Renamed test names from `..._On_NetStandard_...` to `..._On_NonPlatform_...` for TFM-accurate naming.

✅ Validation run
- `dotnet build src/Essentials/src/Essentials.csproj -f net10.0 --no-restore -v q`
- `dotnet build src/Essentials/test/UnitTests/Essentials.UnitTests.csproj --no-restore -v q`
- `dotnet test src/Essentials/test/UnitTests/Essentials.UnitTests.csproj --no-build --filter "FullyQualifiedName~MainThread_Tests"`
  - Passed: 6, Failed: 0

✅ Push
- Commit: `cd2c87a7ab`
- Branch: `dev/redth/fix-34101`

**Comment by PureWeen** [comment]
## 🔍 Multi-Model Code Review — PR #34230

**PR:** Add `MainThread.SetCustomImplementation` for custom platform backends
**Author:** @Redth | **Base:** `main` | **Issue:** #34101
**Reviewers:** 3 independent automated reviewers with adversarial consensus

---

### CI Status: ✅ All checks passing

All `maui-pr` stages (builds, Helix unit tests, integration tests across platforms) are green.

---

### Prior Review Comments

Previous Copilot reviewer and agent feedback has been addressed by the author (atomicity via single `CustomMainThreadCallbacks` object, `PlatformNotSupportedException` on platform TFMs, test naming). This review covers the current state of the diff.

---

## Findings

### 🟡 MODERATE — Branch targeting: new public API on `main` instead of `net11.0`
**Files:** `MainThread.shared.cs`, all 8 `PublicAPI.Unshipped.txt` files
**Consensus:** 3/3 reviewers agree

Per MAUI repo conventions, `main` is for bug fixes without API changes, and `net11.0` is for new features and API changes. This PR introduces a **new public API** (`SetCustomImplementation`) confirmed by the `PublicAPI.Unshipped.txt` additions, but targets `main`. Even though the motivation is a bug fix for custom backends, the vehicle is a new public method.

**Ask:** Should this target `net11.0`, or does the team have explicit sign-off for shipping new API from `main`?

---

### 🟡 MODERATE — `ArgumentNullException` fires before `PlatformNotSupportedException` on platform TFMs
**File:** `src/Essentials/src/MainThread/MainThread.shared.cs` (lines ~37-48)
**Consensus:** 3/3 reviewers agree

On platform TFMs (Android, iOS, etc.), the null guards execute **before** the unconditional `PlatformNotSupportedException` throw:
```csharp
if (isMainThread is null)
    throw new ArgumentNullException(...);  // ← fires first on null input
// ...
throw new PlatformNotSupportedException(...);  // ← never reached if null
```
A caller passing `null` gets `ArgumentNullException` instead of `PlatformNotSupportedException`, implying the method *would* work with valid arguments. This is misleading for feature-detection patterns.

**Suggested fix:** Move the `#if` platform guard **before** the null checks:
```csharp
#if ANDROID || IOS || MACCATALYST || TVOS || WATCHOS || WINDOWS || TIZEN
    throw new PlatformNotSupportedException("...");
#else
    ArgumentNullException.ThrowIfNull(isMainThread);
    ArgumentNullException.ThrowIfNull(beginInvokeOnMainThread);
    customImplementation = new CustomMainThreadCallbacks(isMainThread, beginInvokeOnMainThread);
#endif
```

---

### 🟡 MODERATE — Test parallelism risk with shared static state
**File:** `src/Essentials/test/UnitTests/MainThread_Tests.cs`
**Consensus:** 3/3 reviewers agree

Tests mutate the static `customImplementation` field. xUnit parallelizes across test classes by default. If any other test class in the assembly accesses `MainThread.IsMainThread` or `MainThread.BeginInvokeOnMainThread` concurrently, the static state from these tests will leak, causing non-deterministic failures.

The two "without custom implementation" tests (`IsMainThread_On_NonPlatform_Without_Custom_Implementation`, `BeginInvokeOnMainThread_On_NonPlatform_Without_Custom_Implementation`) are especially vulnerable — they assert the null initial state but don't call `ResetCustomImplementation()` first.

**Suggested fix:** Add `[Collection("MainThread")]` to serialize with any other `MainThread` consumers, or add a setup/teardown that resets state before each test.

---

### 🟡 MODERATE — No public way to clear the custom implementation
**File:** `src/Essentials/src/MainThread/MainThread.shared.cs` — `ResetCustomImplementation()`
**Consensus:** 2/3 reviewers agree

`ResetCustomImplementation()` is `internal`, meaning consumers on custom platforms (the target audience) cannot unregister during backend teardown, hot-reload, or testing in their own projects. The API is one-directional — you can set but never clear without internal access.

**Suggested fix:** Either make `ClearCustomImplementation()` public, or accept `null` in `SetCustomImplementation` as a documented "reset" convention.

---

### 🟢 MINOR — Inconsistent volatile read pattern between the two platform methods
**File:** `src/Essentials/src/MainThread/MainThread.netstandard.cs`
**Consensus:** 2/3 reviewers agree

`PlatformBeginInvokeOnMainThread` uses an explicit local snapshot (`var impl = customImplementation`), while `PlatformIsMainThread` uses `?.` null-conditional. Both are technically safe (C# spec guarantees single evaluation of `?.` receiver), but the inconsistency is a maintenance hazard.

**Suggested fix:** Use the explicit-local pattern in both for consistency.

---

### 🟢 MINOR — User-supplied callback exceptions not documented
**File:** `src/Essentials/src/MainThread/MainThread.shared.cs` — XML docs
**Consensus:** 3/3 reviewers agree

If the custom `isMainThread` or `beginInvokeOnMainThread` delegate throws, the exception propagates directly to callers. Previously, callers only expected `NotImplementedInReferenceAssemblyException`. This contract change should be documented.

**Suggested fix:** Add a `<remarks>` note: *"Exceptions thrown by the provided delegates will propagate to callers of `IsMainThread` and `BeginInvokeOnMainThread`."*

---

### 🟢 MINOR — Silent replacement semantics undocumented
**File:** `src/Essentials/src/MainThread/MainThread.shared.cs`
**Consensus:** 2/3 reviewers agree

Calling `SetCustomImplementation` twice silently replaces the first. In multi-library scenarios, the last caller wins with no indication of overwrite. Consider logging a diagnostic, or at minimum document this is the intended behavior.

---

### 🟢 MINOR — Missing edge-case test coverage
**File:** `src/Essentials/test/UnitTests/MainThread_Tests.cs`
**Consensus:** 3/3 reviewers agree

Three edge cases are untested:
1. **Double-set replacement** — calling `SetCustomImplementation` twice; second should replace first
2. **Callback that throws** — exception should propagate to callers
3. **`InvokeOnMainThreadAsync` with custom impl** — the primary async entry point should flow through the custom implementation

---

### ℹ️ Informational — Discarded findings

| Finding | Source | Reason discarded |
|---------|--------|-----------------|
| Missing `MACOS` in `#if` guard | 1 reviewer (2/3 after adversarial) | **Factually incorrect**: macOS compilation rules are [commented out](https://github.com/dotnet/maui/blob/main/src/Essentials/src/Essentials.csproj#L73) in `Essentials.csproj`. `MACOS` is never defined; macOS is handled through `MACCATALYST`. |
| Inconsistent snapshot race in `BeginInvokeOnMainThread` | 2/3 agreed | Theoretically present but **not practically exploitable**: `SetCustomImplementation` is a configure-once-at-startup API. The double-read across `IsMainThread` → `PlatformBeginInvokeOnMainThread` cannot produce incorrect behavior in any supported usage pattern. |
| Async wrapper sync-throw vs faulted Task | 2/3 agreed | **Pre-existing** structural issue in `InvokeOnMainThreadAsync`, not introduced by this PR. |
| `ResetCustomImplementation` naming | 1/3 only | Below consensus threshold. |

---

### Test Coverage Assessment

The PR adds 6 unit tests covering the core paths: null guards, custom `IsMainThread`, custom `BeginInvokeOnMainThread`, and default exception behavior. Test type (xUnit unit tests) is appropriate for this API-only change. However, 3 edge cases are missing (noted above).

---

### ⚠️ Recommended Action: **Request Changes**

The PR is well-designed architecturally — the immutable `CustomMainThreadCallbacks` class with `volatile` reference is the correct lock-free pattern. However, several issues should be addressed before merge:

1. **Clarify branch targeting** — confirm with maintainers that new API on `main` is intentional
2. **Fix exception ordering** — move `#if` guard before null checks on platform TFMs
3. **Harden test isolation** — add `[Collection]` or setup/teardown to prevent xUnit parallelism flakes
4. **Add missing edge-case tests** — double-set, callback-throws, async path
5. **Document callback exception propagation** — update XML docs

**Comment by Redth** [comment]
Thanks for the review — I addressed the actionable items in the latest updates:

- Retargeted the PR to `net11.0`, since this introduces a new public API.
- Moved the platform guard ahead of the argument validation so platform TFMs consistently throw `PlatformNotSupportedException`.
- Documented the replacement/reset behavior for the custom implementation, and noted that delegate exceptions propagate to callers.
- Expanded and hardened the unit tests to cover static-state cleanup, replacement, clear/reset behavior, exception propagation, and the async path.
- Merged the latest `net11.0` changes and resolved the resulting `PublicAPI.Unshipped.txt` conflicts.

Please take another look when you get a chance.

**Comment by Redth** [comment]
Closing in favour of #35070
